File: src/98. Validate Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

function isValidBSTRecursive(node, min, max) {
  if (node === null) return true
  if (node.val >= max || node.val <= min) return false
  return (
    isValidBSTRecursive(node.right, node.val, max) &&
    isValidBSTRecursive(node.left, min, node.val)
  )
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */

function isValidBST(root) {
  return isValidBSTRecursive(root, -Infinity, Infinity)
}

File: src/98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.valid(root, float('-inf'), float('inf'))
    def valid(self, root: Optional[TreeNode], min: float, max: float):
        if not root:
            return True
        if min >= root.val or root.val >= max:
            return False
        return self.valid(root.left, min, root.val) and self.valid(root.right, root.val, max)